Mathlib/Geometry/Manifold/Instances/Icc.lean

```diff
@@ -0,0 +1,201 @@
+        Function.comp_def, Function.update_self, ModelWithCorners.mk_coe, PartialHomeomorph.mk_coe]
+    rw [Subtype.range_val_subtype]
+    have : ContDiff ℝ n (fun (z : EuclideanSpace ℝ (Fin 1)) ↦ y - z 0) := by fun_prop
+    apply this.contDiffWithinAt.congr_of_eventuallyEq_of_mem; swap
+    · simpa using z.2.2
+    have : {w : EuclideanSpace ℝ (Fin 1) | w 0 < y - x} ∈ 𝓝 (fun i ↦ y - z) := by
+      apply (isOpen_lt (continuous_apply 0) continuous_const).mem_nhds
+      simpa using h.out.trans_le hz
+    filter_upwards [self_mem_nhdsWithin, nhdsWithin_le_nhds this] with w hw h'w
+    rw [max_eq_left hw, max_eq_left]
+    linarith
+
+/-- The projection from `ℝ` to a closed segment is smooth on the segment, in the manifold sense. -/
+lemma contMDiffOn_projIcc :
+    ContMDiffOn 𝓘(ℝ) (𝓡∂ 1) n (Set.projIcc x y h.out.le) (Icc x y) := by
+  intro z hz
+  rw [contMDiffWithinAt_iff]
+  refine ⟨by apply ContinuousAt.continuousWithinAt; fun_prop, ?_⟩
+  simp? says
+    simp only [extChartAt, PartialHomeomorph.extend, Icc_chartedSpaceChartAt,
+      PartialEquiv.coe_trans, ModelWithCorners.toPartialEquiv_coe, PartialHomeomorph.toFun_eq_coe,
+      PartialHomeomorph.refl_partialEquiv, PartialEquiv.refl_source,
+      PartialHomeomorph.singletonChartedSpace_chartAt_eq, modelWithCornersSelf_partialEquiv,
+      PartialEquiv.trans_refl, PartialEquiv.refl_symm, PartialEquiv.refl_coe, CompTriple.comp_eq,
+      preimage_id_eq, id_eq, modelWithCornersSelf_coe, range_id, inter_univ]
+  split_ifs with h'z
+  · simp? [IccLeftChart, Function.comp_def, modelWithCornersEuclideanHalfSpace, projIcc] says
+      simp only [modelWithCornersEuclideanHalfSpace, Fin.isValue, ModelWithCorners.mk_coe,
+        IccLeftChart, PartialHomeomorph.mk_coe, Function.comp_def, projIcc]
+    have : ContDiff ℝ n (fun (w : ℝ) ↦
+        (show EuclideanSpace ℝ (Fin 1) from fun (_ : Fin 1) ↦ w - x)) := by
+      dsimp
+      apply contDiff_euclidean.2 (fun i ↦ by fun_prop)
+    apply this.contDiffWithinAt.congr_of_eventuallyEq_of_mem _ hz
+    filter_upwards [self_mem_nhdsWithin] with w hw
+    ext i
+    simp only [sub_left_inj]
+    rw [max_eq_right, min_eq_right hw.2]
+    simp [hw.1, h.out.le]
+  · simp? [IccRightChart, Function.comp_def, modelWithCornersEuclideanHalfSpace, projIcc] says
+      simp only [modelWithCornersEuclideanHalfSpace, Fin.isValue, ModelWithCorners.mk_coe,
+        IccRightChart, PartialHomeomorph.mk_coe, Function.comp_def, projIcc]
+    have : ContDiff ℝ n (fun (w : ℝ) ↦
+        (show EuclideanSpace ℝ (Fin 1) from fun (_ : Fin 1) ↦ y - w)) := by
+      dsimp
+      apply contDiff_euclidean.2 (fun i ↦ by fun_prop)
+    apply this.contDiffWithinAt.congr_of_eventuallyEq_of_mem _ hz
+    filter_upwards [self_mem_nhdsWithin] with w hw
+    ext i
+    simp only [sub_left_inj]
+    rw [max_eq_right, min_eq_right hw.2]
+    simp [hw.1, h.out.le]
+
+lemma contMDiffOn_comp_projIcc_iff {f : Icc x y → M} :
```

Comment:
Medium-term, this should be proven by saying that the projection of the real line is a smooth submersion. (I'm not actively working on this, but am gradually working on the immersion/embedding analogue. Adding the submersions version shouldn't be too hard.) Can you add a TODO stating this, please?

(And perhaps also one for the other direction, f: M \to Icc is smooth iff the composition i \comp f: M \to R is.)

Reply:
In theory it's true, in practice it will depend on the fine details of the definition because the boundaries may create subtleties, I think.

Reply:
My impression from reading e.g. Roig-Domingues' textbook (on Banach manifolds with boundary and corners) is that these cases should be fine: the correct definition of being an immersion is "there exists a slice chart", and in this case the obvious choices work. It shouldn't be hard to formalise this.